### PR TITLE
feat(azure-iothub): add AAD authentication

### PR DESCRIPTION
**Pull request by zolvarga**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-node/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: common/transport/http/src/rest_api_client.ts [hunk 1]
```diff
@@ -109,7 +113,60 @@
     let httpHeaders: any = headers || {};
     if (this._config.sharedAccessSignature) {
       httpHeaders.Authorization = (typeof(this._config.sharedAccessSignature) === 'string') ? this._config.sharedAccessSignature as string : (this._config.sharedAccessSignature as SharedAccessSignature).extend(anHourFromNow());
+      this.executeBody(requestBody, httpHeaders, headers, method, path, timeout, requestOptions, done);
+    } else if (this._config.tokenCredential) {
+      let accessToken = this.getToken();
+      Promise.resolve(accessToken).then((value) => {
```
Comment (anthonyvercolano) [inline common/transport/http/src/rest_api_client.ts:119]: What happens if this rejects?

Comment (anthonyvercolano) [reply common/transport/http/src/rest_api_client.ts:119]: I'm thinking if this rejects, the program will crash.  If not now, then under node 14?

File: common/transport/http/package.json [hunk 2]
```diff
@@ -19,8 +19,9 @@
     "sinon": "^7.4.1",
     "source-map-support": "^0.5.16",
     "ts-node": "^8.6.2",
-    "tslint": "^6.1.3",
-    "typescript": "3.7.5"
+    "tslint": "^6.1.0",
```
Comment (anthonyvercolano) [inline common/transport/http/package.json:22]: Why did we go backwards in the version of tslint?

File: service/src/client.ts [hunk 3]
```diff
@@ -453,7 +455,8 @@ export class Client extends EventEmitter {
     const config: Client.TransportConfigOptions = {
       host: decodedUri,
       keyName: sas.skn,
-      sharedAccessSignature: sas.toString()
+      sharedAccessSignature: sas.toString(),
```
Comment (anthonyvercolano) [inline service/src/client.ts:458]: Why isn't there a fromToken in here and jobClient?

**Review by anthonyvercolano — CHANGES_REQUESTED** [review]
Let's wait on the job and other entries before we check this in,

File: service/src/client.ts (hunk 3 above)
Comment (zolvarga) [reply service/src/client.ts:458]: Added, fixed

File: common/transport/http/src/rest_api_client.ts (hunk 1 above)
Comment (zolvarga) [reply common/transport/http/src/rest_api_client.ts:119]: Error handling added

File: common/transport/http/package.json (hunk 2 above)
Comment (zolvarga) [reply common/transport/http/package.json]: There is no reason :) Rolled back

**Review by anthonyvercolano — APPROVED** [review]
Finish up testing?  Make sure we squash before we merge

**Pull request merged**